core/src/main/scala/org/apache/spark/scheduler/DecommissionTracker.scala

```diff
@@ -0,0 +1,405 @@
+      executorDecommissionTime = executorDecommissionTimeMs,
+      shuffleDataDecommissionTime = shuffleDataDecommissionTimeMs,
+      terminationTime = terminationTimeMs,
+      reason = reason,
+      state = NodeDecommissionState.DECOMMISSIONING)
+
+    logInfo(s"""Adding decommissioning""" +
+      s""" request : {"node":"$hostname",$nodeDecommissionInfo} """)
+
+    // Add node to the list of decommissioning nodes.
+    decommissionHostnameMap.put(hostname, nodeDecommissionInfo)
+
+    // Schedule executor decommission
+    decommissionThread.schedule(new Runnable {
+      override def run(): Unit = Utils.tryLogNonFatalError {
+        executorDecommission(hostname, nodeDecommissionInfo)
+      }
+    }, executorDecommissionTimeMs - curTimeMs, TimeUnit.MILLISECONDS)
+
+    // Schedule shuffle decommission
+    decommissionThread.schedule(new Runnable {
+      override def run(): Unit = Utils.tryLogNonFatalError {
+        removeShuffleData(hostname, nodeDecommissionInfo)
+      }
+    }, shuffleDataDecommissionTimeMs - curTimeMs, TimeUnit.MILLISECONDS)
+  }
+
+  def removeNodeToDecommission(hostname: String): Unit = synchronized {
+    if (!decommissionHostnameMap.contains(hostname)) {
+      return
+    }
+
+    val nodeDecommissionInfo = decommissionHostnameMap(hostname)
+    logInfo(s"""Removing decommissioning""" +
+      s""" request : {"node":"$hostname",$nodeDecommissionInfo}""")
+    decommissionHostnameMap -= hostname
+  }
+
+  def updateNodeToDecommissionSetTerminate(hostname: String): Unit = synchronized {
+    terminate(hostname)
+  }
+
+  private def executorDecommission(hostname: String,
+                                   nodeDecommissionInfo: NodeDecommissionInfo): Unit = {
+    // Not found, only valid scenario is the nodes
+    // has moved back to running state
+    // Scenario where nodeLoss terminated the node
+    // for the Graceful Decommission node.
+    // If the node is already terminated and hostname is re-used in that scenario
+    // no need to kill the executor on that host
+    if (! decommissionHostnameMap.contains(hostname)) {
+      logInfo(s"""Node $hostname not found in decommisssionTrackerList while""" +
+        """performing executor decommission""")
+      return
+    }
+    // if the terminationTime in the thread is not equal to
+    // terminationTime in decommissionHostnameMap for that
+    // host than Ignore the ExecutorDecommission
+    if (decommissionHostnameMap(hostname).terminationTime
+      != nodeDecommissionInfo.terminationTime) {
+      logInfo(s"Ignoring ExecutorDecommission for hostname ${hostname}," +
+        s" since node is already terminated")
+      return
+    }
+
+    // Kill executor if there still are some running. This call is
+    // async does not wait for response. Otherwise it may cause
+    // deadlock between schedulerBacked (ExecutorAllocationManager)
+    // and this.
+    executorAllocClient.map(_.killExecutorsOnHost(hostname))
```

Comment:
this only works with dynamic allocation then?  Decommissioning is just ignored with dynamic allocation off?  This needs to be documented

Reply:
@tgravescs why do you say that ? executorAllocClient would exist whenever the scheduler backend inherits from CoarseGrainedSchedulerBackend, which happens with almost all scheduling modes including local, master, k8s and yarn etc. Perhaps I am not understanding something. Thanks.